### PR TITLE
[Feature] Receive error event when device manager throws any error

### PR DESCRIPTION
**Pull request by Mohtasim**
## Purpose

* Adding unknown error state for ambiguous error state
* Adding error state propagation for device manager error

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Pull Request Checklist

- [x] Public API changes
- [ ] Verified for cross-platform
- [x] Synced with iOS, Web team
- [ ] Internal review completed
- [ ] UI Changes
  - [ ] Screen captures included
  - [ ] Tested for Light/Dark mode
  - [ ] Tested for screen rotation
  - [ ] Tested on Tablet and small screen device (5")
  - [ ] Include localization changes
- [ ] Tests
  - [x] Memory leak analysis performed
  - [ ] Tested on API 21, 26 and latest 
  - [x] Tested for background mode
  - [x] Unit Tests Included
  - [ ] UI Automated Tests Included

## How to Test

* Open CallingMiddlewareActionHandlerUnitTest class on test package
* run the test for function "callingMiddlewareActionHandler_setupCall_fails_then_dispatchFatalError"

## What to Check
Verify that the following are valid
* Verify that the test is passed or not

## Other Information

* User story: [User Story 2630120](https://skype.visualstudio.com/SPOOL/_workitems/edit/2630120): [Android] As a Developer I want to receive error event when device manager throws any error

### CODE REVIEW

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorHandler.kt [hunk 1]
```diff
@@ -110,11 +112,15 @@ internal class ErrorHandler(
     }
 
     private fun getCallCompositeErrorCode(errorCode: ErrorCode?): CallCompositeErrorCode? {
+        Log.d("Mohtasim", "Error code - " + errorCode.toString())
```
Comment (iaulakh) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorHandler.kt:115]: Log should be removed :)

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorCode.kt [hunk 2]
```diff
@@ -14,6 +14,7 @@ internal class ErrorCode : ExpandableStringEnum<ErrorCode?>() {
         val TURN_CAMERA_OFF_FAILED = fromString("turnCameraOffFailed")
         val TURN_MIC_ON_FAILED = fromString("turnMicOnFailed")
         val TURN_MIC_OFF_FAILED = fromString("turnMicOffFailed")
+        val UNKNOWN_ERROR = fromString("unknownError")
```
Comment (iaulakh) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorCode.kt:17]: do we have discussion with iOs team to sync on this error?

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/models/CallCompositeErrorCode.java [hunk 3]
```diff
@@ -27,6 +27,11 @@ public final class CallCompositeErrorCode extends ExpandableStringEnum<CallCompo
      */
     public static final CallCompositeErrorCode TOKEN_EXPIRED = fromString("tokenExpired");
 
+    /***
```
Comment (iaulakh) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/models/CallCompositeErrorCode.java:30]: It is public API change. We may need review for this. Please discuss with Pavel.

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/LocalParticipantAction.kt [hunk 4]
```diff
@@ -9,6 +9,7 @@ import com.azure.android.communication.ui.calling.redux.state.AudioOperationalSt
 import com.azure.android.communication.ui.calling.redux.state.CameraDeviceSelectionStatus
 
 internal sealed class LocalParticipantAction : Action {
+    class DeviceManagerFetchFailed(val error: CallCompositeError) : LocalParticipantAction()
```
Comment (iaulakh) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/LocalParticipantAction.kt:12]: would it be possible to indicate the cause of Fetch failed? like DeviceErrorType

**Review by iaulakh — CHANGES_REQUESTED** [review]

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorCode.kt (hunk 2 above)
Comment (Mohtasim) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorCode.kt:17]: Yes. iOS team and we are on sync with introducing the Unknown Error. 
Basically, we will use this error whenever the composite runs to an ambiguous state which happens due to any system related issue of the device such as this type of error (getting error on device manager)

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorHandler.kt (hunk 1 above)
Comment (Mohtasim) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/error/ErrorHandler.kt:115]: :o My bad. Thanks for the catch :)

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/models/CallCompositeErrorCode.java (hunk 3 above)
Comment (Mohtasim) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/models/CallCompositeErrorCode.java:30]: We (android & iOS) had plans for review but we got the response that we don't need to sit for review to add only an error code.

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/LocalParticipantAction.kt (hunk 4 above)
Comment (Mohtasim) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/LocalParticipantAction.kt:12]: It is documented in the CallCompositeErrorCode class about the unknowError code

**Review by ShaunaSong — DISMISSED** [review]

**Review by ShaunaSong — DISMISSED** [review]

**Review by ShaunaSong — APPROVED** [review]

**Review by iaulakh — APPROVED** [review]

**Pull request merged**